DirectEmployers/MyJobs
# [PD-3031] user locations from ip

WARNING: A tiny bit of the logic surrounding the database might change depending on some on exactly what version of the database we decide to go with. This also does not add location to reporting. Neither of these will have a huge impact on the logic in this pull request. Both of those will be handled, as needed, in a different pull request.


## I highly recommend that everyone adds the following line to dev_settings.py
`MMDB_PATH = 'db.mmdb'`
This prevents constantly reloading the database every time you create a new docker container (so use runserver, open a shell, etc).



## What's this PR do?
Given the IP address of a user, maps the IP to a location.

## Where should the reviewer start?
Most of the new logic starts in `models.py`. More specifically, `UserLocationBaseModel` will be the entry point, while the code in `user_locations.py` does the heavy lifting.

`UserLocationBaseModel` has custom logic for `save` and `__init__` in order to ensure locations are added. Is there any other case I missed besides those two?

The logic in `user_locations.init_db` and `user_locations.reload_db` is meant to ensure the database can be loaded/updated dynamically and, more importantly, without managing the database in source control.

## How should this be manually tested?
Run the necessary migrations and create some new `JobView` objects with IP addresses. Confirm that the user location is getting added.
